Replace notification with confirm dialog for user deletion

sharidas opened the pull request:
Replace notification with confirm dialog for user
deletion from UI.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
fixes https://github.com/owncloud/enterprise/issues/2755

PVince81 commented:
@sharidas the code cleanup looks incomplete, please grep for "undo"

sharidas commented:
- [ ] Is the undo code removed completely?

sharidas commented:
Current state of the PR:
- Remvoved "undo" from the code.
- Tested the UI, by adding user and deleting user
- Tested by adding user to a group(s) and deleting the user

Grepping "undo" is now restricted to l10n folder and `deleteHandlerSpec.js`. For `deleteHandlerSpec.js`, the undo is associated with `undoCallback`.

PVince81 commented:
@sharidas ready for review ? the subject still says WIP. please ping when ready...

sharidas commented:
@PVince81 kindly review this pr.

PVince81 commented:
Code looks fine so far.

Please have a look at the acceptance tests

sharidas commented:
There is one thing that I have noticed while adjusting the acceptance test. While replacing undo for users, wouldn't it be better to have the same for groups too? Because they do share the same code. Here is a check list that needs to be addressed:

- [ ] Should we replace notification for group deletion with confirm dialog?

PVince81 commented:
@sharidas Let's do it for groups as well, for consistency. Thanks for spotting!

sharidas commented:
Current state of the PR:
- [x] Replace group deletion from notification to confirmation ( same as user deletion )
- [x] Updated the acceptance test accordingly

sharidas commented:
@phil-davis @individual-it  kindly review my changes made for acceptance tests. 
Just to add the context of this PR:
- Replace the notification displayed when user is deleted from webUI to confirmation dialog.
- Replace the notification displayed when group is deleted from webUI to confirmation dialog.

phil-davis commented on js/groups.js:
```diff
@@ -269,22 +269,25 @@ GroupList = {
-			GroupDeleteHandler.deleteEntry();
+			var groupName = encodeURIComponent(GroupList.getElementGID(this)).toString();
+			OC.dialogs.confirm(
+				t('user_management', 'You are about to delete a group. This action can\'t be undone and is permanent. Are you sure that you like to permanently delete {groupName}?', {groupName: decodeURIComponent(groupName)}),
```
I suggest
s/that you like to/that you want to/

phil-davis commented on js/users.js:
```diff
@@ -377,13 +377,22 @@ var UserList = {
 			var uid = UserList.getUID(this);
-			UserDeleteHandler.mark(uid);
+			OC.dialogs.confirm(
+				t('user_management', 'You are about to delete a user. This action can\'t be undone and is permanent. All user data, files and shares will be deleted. Are you sure that you like to permanently delete {userName}?', {userName: uid}),
```
I suggest
s/that you like to/that you want to/

phil-davis commented on js/users.js:
```diff
@@ -377,13 +377,22 @@ var UserList = {
 
 		//delete a marked user when leaving the page
-		$(window).on('beforeunload', function () {
+		/*$(window).on('beforeunload', function () {
```
Delete commented-out code?

phil-davis commented on tests/acceptance/features/bootstrap/WebUIUsersContext.php:
```diff
@@ -290,14 +290,26 @@ public function theDisabledUserTriesToLogin($username, $password) {
 
 	/**
-	 * @When the administrator deletes user :username using the webUI
+	 * @When the administrator deletes the user :username using the webUI and confirms clicking yes
```
s/the user/user/

When we mention a specific user name, we do not put "the" in front of it. Both read OK in English. We chose the short form as a convention.

phil-davis replied:
"and confirms the deletion using the webUI"

phil-davis commented on tests/acceptance/features/bootstrap/WebUIUsersContext.php:
```diff
@@ -153,19 +153,19 @@ public function theAdminCreatesAUserUsingWithoutAPasswordTheWebUI(
 	/**
 	 *
-	 * @When the administrator deletes the group named :name using the webUI
+	 * @When the administrator deletes the group named :name using the webUI and confirms clicking yes
```
"and confirms the deletion using the webUI"

phil-davis commented on tests/acceptance/features/bootstrap/WebUIUsersContext.php:
```diff
@@ -153,19 +153,19 @@ public function theAdminCreatesAUserUsingWithoutAPasswordTheWebUI(
 
 	/**
-	 * @When the administrator deletes these groups using the webUI:
+	 * @When the administrator deletes these groups and confrims clicking yes using the webUI:
```
"and confirms the deletion using the webUI"

(note the typo in "confrims" needs to be fixed while editing)

phil-davis commented on tests/acceptance/features/bootstrap/WebUIUsersContext.php:
```diff
@@ -290,14 +290,26 @@ public function theDisabledUserTriesToLogin($username, $password) {
+	}
+
+	/**
+	 * @When the administrator deletes the user :username using the webUI and confirms clicking no
```
"and cancels the deletion using the webUI"

phil-davis commented on tests/acceptance/features/bootstrap/WebUIUsersContext.php:
```diff
@@ -290,14 +290,26 @@ public function theDisabledUserTriesToLogin($username, $password) {
+	 *
+	 * @return void
+	 */
+	public function theAdminstratorDoesNotDeleteTheUser($username) {
```
s/Adminstrator/Administrator/

phil-davis commented on tests/acceptance/features/lib/UserPageElement/GroupList.php:
```diff
@@ -101,6 +106,22 @@ public function deleteGroup($name) {
+			throw new ElementNotFoundException(
+				__METHOD__ .
+				" xpath $xpathSelector " .
+				"could not find quota element"
```
"could not find delete confirm button"

phil-davis commented on tests/acceptance/features/lib/UsersPage.php:
```diff
@@ -549,5 +555,22 @@ public function deleteUser($username) {
+			throw new ElementNotFoundException(
+				__METHOD__ .
+				" xpath $xpathSelector " .
+				"could not find quota element"
```
"could not find delete confirm button"

sharidas commented:
@phil-davis I have addressed your comments in my PR
- [x] rephrased `and confirms clicking yes` with `and confirms the deletion using the webUI`
- [x] rephrased `and confirms clicking no` with `and cancels the deletion using the webUI`
- [x] rephrased the string in the confirmation dialog to `you want to`
- [x] Added simple scenarios to cancel user/group deletion
- [x] Removed the dead code ( commented code ) 

Hence kindly re-review my change.

phil-davis commented on tests/acceptance/features/bootstrap/WebUIUsersContext.php:
```diff
@@ -290,14 +316,26 @@ public function theDisabledUserTriesToLogin($username, $password) {
 
 	/**
-	 * @When the administrator deletes user :username using the webUI
+	 * @When the administrator deletes user :username using the webUI and confirms the deletion using the webUI
```
I wasn't thinking about repeating ``using the webUI`` but I guess it makes sense, because we have to be clear that the step does both things ``using the webUI``

sharidas replied:
> I wasn't thinking about repeating `using the webUI` but I guess it makes sense, because we have to be clear that the step does both things `using the webUI`

Yah I was trying to make it explicit.

phil-davis commented on tests/acceptance/features/bootstrap/WebUIUsersContext.php:
```diff
@@ -290,14 +316,26 @@ public function theDisabledUserTriesToLogin($username, $password) {
+	}
+
+	/**
+	 * @When the administrator deletes the user :username using the webUI and cancels the deletion using the webUI
```
s/deletes the user/deletes user/

phil-davis commented on tests/acceptance/features/webUIManageUsersGroups/manageGroups.feature:
```diff
@@ -115,3 +115,45 @@ Feature: manage groups
+    And the administrator reloads the users page
+    Then group "space group" should exist
+
+  Scenario: cancel deletion of a groups
```
cancel deletion of groups

sharidas commented:
@phil-davis  I have updated the changes requested. Let me know if you are ok with this change set.

PVince81 commented on js/users.js:
```diff
@@ -377,12 +377,16 @@ var UserList = {
-		$(window).on('beforeunload', function () {
-			UserDeleteHandler.deleteEntry();
+			OC.dialogs.confirm(
+				t('user_management', 'You are about to delete a user. This action can\'t be undone and is permanent. All user data, files and shares will be deleted. Are you sure that you want to permanently delete {userName}?', {userName: uid}),
```
Add space between username and question mark.

As far as I remember in English we always have a space there.

phil-davis replied:
```
Are you sure that you want to permanently delete user123?
```
is fine.

There is no space before a ``?``

sharidas replied:
@PVince81 @phil-davis 

For instance lets take a look at https://www.gnu.org/philosophy/free-sw.html. 

- The heading `What is free software?` does not have space before `?`. 
- Another instance `Have a question about free software licensing not answered here?` in the same page.

Let me know if we agree on no space before `?` :)

phil-davis replied:
@PVince81 already clicked the approve button. He will be happy! I will mark this discussion resolved.

phil-davis commented:
@sharidas this repo insists on the PR being rebased to the tip of master before allowing merge. So I guess you will have to rebase it again.

@PVince81 is this repo set like that for a reason? Or just an accident?

sharidas commented:
> @sharidas this repo insists on the PR being rebased to the tip of master before allowing merge. So I guess you will have to rebase it again.
> 

@phil-davis  Thanks. I have clicked on the `Updated branch` and the drone is running CI.

sharidas commented:
Backport PR to stable10 branch: https://github.com/owncloud/core/pull/33626

PVince81 commented on js/groups.js:
```diff
@@ -269,22 +269,25 @@ GroupList = {
-			GroupDeleteHandler.deleteEntry();
+			var groupName = encodeURIComponent(GroupList.getElementGID(this)).toString();
+			OC.dialogs.confirm(
+				t('user_management', 'You are about to delete a group. This action can\'t be undone and is permanent. Are you sure that you want to permanently delete {groupName}?', {groupName: decodeURIComponent(groupName)}),
```
can you add a space before the question mark in a subsequent PR ? @sharidas

phil-davis replied:
@settermjd can you advise here please?
I have never heard of putting a space before a question mark.

PVince81 replied:
never mind... I just checked the l10n files and apparently there is never a space.

so I've been doing this wrong all these years ?
in some languages you need a space there

PVince81 replied:
never mind... I just checked the l10n files and apparently there is never a space.

so I've been doing this wrong all these years ?
in some languages you need a space there

this one confirms it: https://english.stackexchange.com/a/4646

apparently the fact that I'm french has biased me: http://grammaire.reverso.net/5_1_10_les_espaces_et_la_ponctuation.shtml (see example sentences)

seems German also doesn't have the space: https://www.duden.de/sprachwissen/rechtschreibregeln/fragezeichen

settermjd replied:
Ah, English, the inconsistent beast that it is.
